web/src/components/vis/mixins/axisPlot.js

```diff
@@ -11,46 +10,72 @@ function labelAxis(label, msg, xFunc, yFunc, rot) {
 }
 
 export const axisPlot = {
-  methods: {
-    axisPlot({
-      margin, xrange, yrange, xlabel, ylabel, duration,
-    }) {
-      // Collect necessary props.
-      const {
-        width,
-        height,
-      } = this.$props;
-
-      // Grab the root SVG element.
-      const svg = select(this.$refs.svg);
-      this.svg = svg;
-
-      // Compute the size of the data rectangle.
-      this.margin = margin;
-
-      const dwidth = width - margin.left - margin.right;
-      this.dwidth = dwidth;
-
-      const dheight = height - margin.top - margin.bottom;
-      this.dheight = dheight;
-
-      // Create X and Y scales.
-      const scaleX = scaleLinear()
+  props: {
+    width: {
+      type: Number,
+      required: true,
+    },
+    height: {
+      type: Number,
+      required: true,
+    },
+  },
+  data() {
+    return {
+      axisPlotInitialized: false,
+      margin: {
+        top: 20,
+        right: 0,
+        bottom: 50,
+        left: 50,
+      },
+      xrange: [-1.2, 1.2],
+      yrange: [-1.2, 1.2],
+      xlabel: 'x axis',
+      ylabel: 'y axis',
+      duration: 500,
+    };
+  },
+  computed: {
+    dwidth() {
+      const { width, margin } = this;
+      return width - margin.left - margin.right;
+    },
+    dheight() {
+      const { height, margin } = this;
+      return height - margin.top - margin.bottom;
+    },
+    scaleX() {
+      const { xrange, dwidth } = this;
+      return scaleLinear()
         .domain(xrange)
         .range([0, dwidth]);
-      this.scaleX = scaleX;
-
-      const scaleY = scaleLinear()
+    },
+    scaleY() {
+      const { yrange, dheight } = this;
+      return scaleLinear()
         .domain(yrange)
         .range([dheight, 0]);
-      this.scaleY = scaleY;
-
-      // Create X and Y axis objects.
-      const axisX = axisBottom(scaleX);
-      this.axisX = axisX;
+    },
+    axisX() {
+      return axisBottom(this.scaleX);
+    },
+    axisY() {
+      return axisLeft(this.scaleY);
+    },
```

Comment:
So, I do have an itch about `axisX()` and `axisY()` as updated (and, to a lesser extent, `scaleX()` and `scaleY()` above as well): these are D3 functions that return what is intended to be a persistent object for use in the rest of the code. In particular, they are *mutable* objects, which means that users of this mixin who are accustomed to the D3 way of doing things might mutate them in various ways, then find that the mutation semantics are weirdly ephemeral (or, perhaps the way the mixin mechanism works means that users of the mixin do not have access to these values; it's not super clear to me). However, this problem is mostly academic for now, because *my* current usage of these values doesn't rely on mutating them.

The other problem is of performance: it feels weird to keep reconstructing these as needed, even though I don't know how heavy the creation process actually is. Likely it just won't matter but this concern also interacts with the mutation-semantics problem above to produce a decidedly non-D3 way of treating these values.

This exposes the fundamental divide between Vue's reactivity and D3's OOPness. I wonder if there's a way to have both: can this mixin be allowed to memoize the values of these objects? I believe that would alleviate both of my general concerns, at the cost of (1) creating extra logic for memoization and (2) creating a risk that this logic will violate Vue's assumptions that underpin the reactivity system.

Let me know what you think of this.

Reply:
That argument makes sense to me.  If that's the case, I think this mixin could be made safer by leaning on the lifecycle.

1. Set `axisX, axisY, scaleX, scaleY` as null in mixin data
2. Override whatever data you want from the component 
3. Inside the `mounted()` hook of the mixin (which will be called before the instantiated component) use the data + computed properties to set those 4 mutables, so they're ready by the time you call `axisPlot`.  In `axisPlot`, check that they're ready and throw some error if they aren't.

... interrupted by chat.  Will continue thoughts soon.

Reply:
This turns out to be harder than I thought. I am going to leave it as-is for now, and if it causes trouble in the future we can revisit.